### PR TITLE
Fix S3 ObjectStore proxy option

### PR DESCRIPTION
**Pull request by maxbes**
The `proxy` option in S3 object store does not current work

```
  'objectstore' => 
  array (
    'class' => '\\OC\\Files\\ObjectStore\\S3',
    'arguments' => 
    array (
    # other_arguments
      'proxy' => 'localhost:3128',
    ),
```
tries to set an unknown/deprecated "request.options" key when constructing a S3Client. The [correct one](https://docs.aws.amazon.com/aws-sdk-php/v3/api/class-Aws.AwsClient.html#___construct) is "http"

Additionally, S3ObjectTrait::readObject uses fopen() directly, so object store proxy options must also be passed to it

### CODE REVIEW

**Review by rullzer — APPROVED** [review]
I can't test but it looks sane

File: lib/private/Files/ObjectStore/S3ObjectTrait.php [hunk 1]
```diff
@@ -70,6 +70,11 @@ public function readObject($urn) {
 				],
 			];
 
+			if ($this->getProxy()) {
+				$opts['http']['proxy'] = $this->getProxy();
+				$opts['http']['request_fulluri'] = true;
```
Comment (juliushaertl) [inline lib/private/Files/ObjectStore/S3ObjectTrait.php:75]: Is there a reason this is only required on the readObject? I guess it depends on the proxy but might make sense to set if for all requests that would use the proxy

Comment (orandev) [reply lib/private/Files/ObjectStore/S3ObjectTrait.php:75]: The other requests use "$this->getConnection()", so they get the proxy options fixed by this PR.
readObject is using fopen directly and thus wasn't using the proxy options from the S3 Connection object.

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php [hunk 2]
```diff
@@ -123,7 +131,7 @@ public function getConnection() {
 			'csm' => false,
 		];
 		if (isset($this->params['proxy'])) {
-			$options['request.options'] = ['proxy' => $this->params['proxy']];
+			$options['http'] = [ 'proxy' => $this->params['proxy'] ];
```
Comment (juliushaertl) [inline lib/private/Files/ObjectStore/S3ConnectionTrait.php:134]: Would also make sense to just use `$this->getProxy()` here

Comment (maxbes) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php:134]: fixed

Comment (juliushaertl) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php:134]:
```suggestion
			$options['http'] = [ 'proxy' => $this->getProxy() ];
```

Another one :wink:

Comment (maxbes) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Yes of course, my bad, I fixed this one too

**Review by juliusknorr — APPROVED** [review]

**Comment by maxbes** [comment]
Sorry for the delay in fixing this, I did not realize the comment on getProxy() was a blocker. Any chance this PR could get in NC22?

**Comment by juliushaertl** [comment]
Failures unrelated

**Comment by orandev** [comment]
Hello @juliushaertl 
I see no PR for the backport to NC 22 and I don't see the code of the PR in NC 22.1.1, is it normal?

**Comment by juliushaertl** [comment]
Thanks for the hint, seems the backport bot was not picking that up. Let me retry.

**Pull request merged**